feat: allow for fine grained invalidation of search params (#11066)
sveltejs/kit: documentation/docs/20-core-concepts/20-load.md

Dependency tracking does not apply _after_ the `load` function has returned — for example, accessing `params.x` inside a nested [promise](#streaming-with-promises) will not cause the function to rerun when `params.x` changes. (Don't worry, you'll get a warning in development if you accidentally do this.) Instead, access the parameter in the main body of your `load` function.

> When setting the `fineGrainedSearchParamsInvalidation` option to `true`, accessing a query parameter is tracked independently from the rest of the url. For example, accessing `event.url.searchParams.get("query")` inside a `load` function will make that `load` function rerun only when the `query` search param changes: Navigating from `/search?query=svelte&page=1` to `/search?query=svelte&page=2` will not rerun the load function.

Reviewer: I think it would be good to mention here that the option is set in `svelte.config.js`. Otherwise it's not clear where to actually set this option.

Author: I agree, will fix soon

Author: Fixed now, thanks

> Also accessing one single query parameter is tracked independently from the rest of the url. This means that accessing `event.url.searchParams.get("query")` inside a `load` function will make that load function rerun only when the `query` search param changes. For example navigating from `/search?query=svelte&page=1` to `/search?query=svelte&page=2` will not rerun a load function that access `event.url.searchParams.get("query")` but not `event.url.searchParams.get("page")`.

Reviewer 2: I think this has more context when inserted after the statement on line 601.

> ### When do load functions rerun?
> To summarize, a `load` function will rerun in the following situations:
> - It references a property of `url` (such as `url.pathname` or `url.search`) whose value has changed. Properties in `request.url` are _not_ tracked
> - _It references a URL query parameter, etc.._

Author: I feel like it would be fitting in both places...I will definitely correct that line too, thanks

Author: I've added the line after 601 but for now i've not removed the previews one

Reviewer 2: Awesome. Thanks for taking on this PR. It’s looking real good :)

Author: I'm glad to be able to help and i also need this sooooo bad so it's a win-win 😄

> Also accessing one single query parameter is tracked independently from the rest of the url. This means that accessing `event.url.searchParams.get("query")` inside a `load` function will make that load function rerun only when the `query` search param changes. For example navigating from `/search?query=svelte&page=1` to `/search?query=svelte&page=2` will not rerun a load function that access `event.url.searchParams.get("query")` but not `event.url.searchParams.get("page")`.

Reviewer 3: ```suggestion
When setting the `fineGrainedSearchParamsInvalidation` option to `true`, accessing a query parameter is tracked independently from the rest of the url. For example, accessing `event.url.searchParams.get("query")` inside a `load` function will make that `load` function rerun only when the `query` search param changes: Navigating from `/search?query=svelte&page=1` to `/search?query=svelte&page=2` will not rerun the load function.
```

- It references a property of `params` whose value has changed
- It references a property of `url` (such as `url.pathname` or `url.search`) whose value has changed. Properties in `request.url` are _not_ tracked

> - It calls `url.searchParams.get`, `url.searchParams.getAll` or `url.searchParams.has` and the specific search param passed to those functions changes. Accessing other properties of searchParams will have the same effect as accessing `url.search`

Reviewer 3: ```suggestion
- It calls `url.searchParams.get`, `url.searchParams.getAll` or `url.searchParams.has` and the specific search param passed to those functions changes. Accessing other properties of searchParams will have the same effect as accessing `url.search`. When `fineGrainedSearchParamsInvalidation` is `false`, accessing _any_ search param will cause the load function to rerun.
```
